cri.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cri
+
+import (
+	"os"
+
+	"github.com/containerd/containerd/plugin"
+	"github.com/docker/docker/pkg/reexec"
+	"github.com/golang/glog"
+
+	"github.com/kubernetes-incubator/cri-containerd/cmd/cri-containerd/options"
+	"github.com/kubernetes-incubator/cri-containerd/pkg/server"
+)
+
+// Register CRI service plugin
+func init() {
+	// TODO(random-liu): Figure out whether we could better handle reexec
+	// or get rid of it.
+	if reexec.Init() {
+		os.Exit(0)
+	}
+	plugin.Register(&plugin.Registration{
+		// In fact, cri is not strictly a GRPC plugin now.
+		Type: plugin.GRPCPlugin,
+		ID:   "cri",
+		Requires: []plugin.Type{
+			plugin.RuntimePlugin,
+			plugin.SnapshotPlugin,
+			plugin.TaskMonitorPlugin,
+			plugin.DiffPlugin,
+			plugin.MetadataPlugin,
+			plugin.ContentPlugin,
+			plugin.GCPlugin,
+		},
+		InitFn: initCRIService,
+	})
+}
+
+func initCRIService(_ *plugin.InitContext) (interface{}, error) {
+	// TODO(random-liu): Support Config through Registration.Config.
+	// TODO(random-liu): Validate the configuration.
+	// TODO(random-liu): Leverage other fields in InitContext, such as Root.
+	// TODO(random-liu): Register GRPC service onto containerd GRPC server.
+	// TODO(random-liu): Separate cri plugin config from cri-containerd server config,
+	// because many options only make sense to cri-containerd server.
```

Comment:
would like to talk about the desire to keep a cri-containerd standalone option alive vs. forking and dropping the server in the plugin fork.

Reply:
Removing a grpc is better, but we won't use the containerd api. It is difficult to select.:)
What is the final decision?